group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ServerSideStickyRangeAssignor.java

```diff
@@ -0,0 +1,268 @@
+                    membersWithNewAssignmentPerTopic.computeIfAbsent(memberId, k -> new HashMap<>()).computeIfAbsent(topicId, k -> new HashSet<>()).add(currentAssignmentListForTopic.get(minRequiredQuota));
+                } else {
+                    // 3) If remaining = 0 it has min req partitions but there is scope for getting an extra partition later on, so it is a potentialUnfilledConsumer.
+                    // 4) If remaining > 0 it doesn't even have the min required partitions, and it definitely is unfilled so it should be added to potentialUnfilledConsumers.
+                    Pair<String, Integer> newPair = new Pair<>(memberId, remaining);
+                    potentiallyUnfilledConsumers.add(newPair);
+                }
+            }
+
+            // Step 3
+            // Iterate through potentially unfilled consumers and if remaining > 0 after considering the extra partitions assignment, add to the unfilled list.
+            for (Pair<String, Integer> pair : potentiallyUnfilledConsumers) {
+                String memberId = pair.getFirst();
+                Integer remaining = pair.getSecond();
+                if (numConsumersWithExtraPartition > 0) {
+                    remaining++;
+                    numConsumersWithExtraPartition--;
+                }
+                if (remaining > 0) {
+                    Pair<String, Integer> newPair = new Pair<>(memberId, remaining);
+                    putList(unfilledConsumersPerTopic, topicId, newPair);
+                }
+            }
+        }
+
+        // Step 4
+        // Find the difference between the total partitions per topic and the already assigned sticky partitions for the topic to get available partitions.
+        Map<Uuid, List<Integer>> availablePartitionsPerTopic = getAvailablePartitionsPerTopic(assignmentSpec, assignedStickyPartitionsPerTopic);
+
+        // Step 5
+        // Iterate through the unfilled consumers list and assign remaining number of partitions from the availablePartitions list
+        for (Map.Entry<Uuid, List<Pair<String, Integer>>> unfilledEntry : unfilledConsumersPerTopic.entrySet()) {
+            Uuid topicId = unfilledEntry.getKey();
+            List<Pair<String, Integer>> unfilledConsumersForTopic = unfilledEntry.getValue();
+
+            for (Pair<String, Integer> stringIntegerPair : unfilledConsumersForTopic) {
+                String memberId = stringIntegerPair.getFirst();
+                int remaining = stringIntegerPair.getSecond();
+                // assign the first few partitions from the list and then delete them from the availablePartitions list for this topic
+                List<Integer> subList = availablePartitionsPerTopic.get(topicId).subList(0, remaining);
+                membersWithNewAssignmentPerTopic.computeIfAbsent(memberId, k -> new HashMap<>()).computeIfAbsent(topicId, k -> new HashSet<>()).addAll(subList);
+                availablePartitionsPerTopic.get(topicId).removeAll(subList);
```

Comment:
i think we can make this more efficient. removing all elements from one list with another will take O(n^2) assuming both have same size.

Reply:
okay yeah makes sense, I'll try to do it with a counter